### PR TITLE
Periodic update network

### PR DESCRIPTION
**Pull request by DmytroLinkin**
Some failed pods in add periodic update may become running in the process #65

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -93,6 +93,11 @@ data:
   GUID_POOL_RANGE_END: "02:FF:FF:FF:FF:FF:FF:FF" # The last guid in the pool
 ```
 
+### Multus configuration
+
+Note: If "kubeconfig" field is missed in 00-multus.conf then every pod creation will ignore network
```
Comment (adrianchiris) [inline README.md:98]: > Note: If "kubeconfig" field is missed in 00-multus.conf

Note: If `kubeconfig` field is missing in multus CNI configuration ...

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -55,6 +55,14 @@ type networksMap struct {
 	theMap map[types.UID][]*v1.NetworkSelectionElement
 }
 
+// Exponential backoff ~26 sec + 6 * <api call time>
+// NOTE: k8s client has built in exponential backoff, which ib-kubernetes don't use.
+// In case of increased backoff time might be KUBE_CLIENT_BACKOFF_BASE & KUBE_CLIENT_BACKOFF_DURATION
```
Comment (adrianchiris) [inline pkg/daemon/daemon.go:60]: Its not clear from the comment what the maximal backoff time would be.
is it the maximal value between the two ? or something else ? as we are performing exponential backoff of an exponential backoff method ?

File: pkg/daemon/daemon.go [hunk 3]
```diff
@@ -107,8 +115,16 @@ func NewDaemon() (Daemon, error) {
 		return nil, err
 	}
 
-	if err := smClient.Validate(); err != nil {
-		return nil, err
+	// Try to validate if subnet manager is reachable in backoff loop
+	var validateErr error
```
Comment (adrianchiris) [inline pkg/daemon/daemon.go:119]: i think for the initial startup of the daemon, we don't need to do exponential backoff WDYT ?

File: pkg/daemon/daemon.go [hunk 4]
```diff
@@ -155,9 +171,19 @@ func (d *daemon) getIbSriovNetwork(networkID string) (string, *utils.IbSriovCniS
...
+	if err = wait.ExponentialBackoff(backoffValues, func() (bool, error) {
+		netAttInfo, err = d.kubeClient.GetNetworkAttachmentDefinition(networkNamespace, networkName)
+		if err != nil {
+			log.Warn().Msgf("failed to get networkName attachment %s with error %v",
+				networkName, err)
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		return "", nil, errcode.Errorf(ErrGetNetAttachDef, "failed to get networkName attachment %s",
```
Comment (adrianchiris) [inline pkg/daemon/daemon.go:185]: should we also add the last error to the returned error for completeness ?

File: pkg/daemon/daemon.go [hunk 5]
```diff
@@ -277,11 +303,20 @@ func (d *daemon) updatePodNetworkAnnotation(pi *podNetworkInfo, removedList *[]n
...
+			if kerrors.IsNotFound(err) {
+				return false, err
+			}
+			log.Warn().Msgf("failed to update pod annotations with err: %v", err)
+			return false, nil
 		}
 
+		return true, nil
+	}); err != nil {
+		log.Error().Msgf("failed to update pod annotations")
```
Comment (adrianchiris) [inline pkg/daemon/daemon.go:319]: should we also add the last error to the returned error for completeness ?

File: pkg/daemon/daemon.go [hunk 6]
```diff
@@ -332,7 +370,8 @@ func (d *daemon) AddPeriodicUpdate() {
 		var failedPods []*kapi.Pod
 		for _, pod := range pods {
 			log.Debug().Msgf("pod namespace %s name %s", pod.Namespace, pod.Name)
-			pi, err := getPodNetworkInfo(networkName, pod, netMap)
+			var pi *podNetworkInfo
+			pi, err = getPodNetworkInfo(networkName, pod, netMap)
```
Comment (adrianchiris) [inline pkg/daemon/daemon.go:372]: why was this change needed ?

File: pkg/daemon/daemon.go [hunk 7]
```diff
@@ -352,15 +391,23 @@ func (d *daemon) AddPeriodicUpdate() {
 
 		// Get configured PKEY for network and add the relevant POD GUIDs as members of the PKey via Subnet Manager
 		if ibCniSpec.PKey != "" && len(guidList) != 0 {
-			pKey, err := utils.ParsePKey(ibCniSpec.PKey)
+			var pKey int
+			pKey, err = utils.ParsePKey(ibCniSpec.PKey)
```
Comment (adrianchiris) [inline pkg/daemon/daemon.go:389]: same question as in L#374

File: pkg/daemon/daemon.go [hunk 8]
```diff
@@ -502,10 +574,19 @@ func (d *daemon) DeletePeriodicUpdate() {
 //  initPool check the guids that are already allocated by the running pods
 func (d *daemon) initPool() error {
 	log.Info().Msg("Initializing GUID pool.")
-	pods, err := d.kubeClient.GetPods(kapi.NamespaceAll)
-	if err != nil {
-		err = fmt.Errorf("failed to get pods from kubernetes: %v", err)
-		log.Err(err)
+
+	// Try to get pod list from k8s client in backoff loop
```
Comment (adrianchiris) [inline pkg/daemon/daemon.go:554]: i think for the initial startup of the daemon, we don't need to do exponential backoff WDYT ?

File: pkg/daemon/daemon.go [hunk 9]
```diff
@@ -352,46 +385,54 @@ func (d *daemon) AddPeriodicUpdate() {
...
 		// Update annotations for PODs that finished the previous steps successfully
 		var removedGUIDList []net.HardwareAddr
 		for _, pi := range passedPods {
-			err := d.updatePodNetworkAnnotation(pi, &removedGUIDList)
-			if err != nil {
-				failedPods = append(failedPods, pi.pod)
-				log.Error().Msgf("%v", err)
-				continue
-			}
+			d.updatePodNetworkAnnotation(pi, &removedGUIDList)
```
Comment (adrianchiris) [inline pkg/daemon/daemon.go:412]: can we keep the original updatePodNetworkAnnotation signature and return error, then log it here ? it seems clearer to me that way so the the main flow is "aware" of the failure

File: pkg/daemon/daemon.go [hunk 10]
```diff
@@ -319,29 +353,28 @@ func (d *daemon) AddPeriodicUpdate() {
...
 		if err != nil {
-			if errcode.GetCode(err) == ErrNotIBSriovNetwork {
-				// Not an IB SR-IOV network
+			code := errcode.GetCode(err)
+			if code == ErrNotIBSriovNetwork || code == ErrGetNetAttachDef {
+				// Not an IB SR-IOV network or failed to get network attachment definition
 				addMap.UnSafeRemove(networkID)
+				log.Error().Msgf("droping network: %v", err)
+			} else {
+				log.Debug().Msgf("skipping network: %v", err)
```
Comment (adrianchiris) [inline pkg/daemon/daemon.go:362]: in this case we will retry this network in the next periodic update call right ?

File: pkg/daemon/daemon.go (hunk 3 above)
Comment (DmytroLinkin) [reply pkg/daemon/daemon.go:119]: Not sure. What if daemon initialized before manager was configured/becomes reachable?

File: pkg/daemon/daemon.go (hunk 4 above)
Comment (DmytroLinkin) [reply pkg/daemon/daemon.go:185]: We discussed it week(?) ago and decided that warning inside backoff loop is enough.

File: pkg/daemon/daemon.go (hunk 5 above)
Comment (DmytroLinkin) [reply pkg/daemon/daemon.go:319]: same answer as previous

File: pkg/daemon/daemon.go (hunk 6 above)
Comment (DmytroLinkin) [reply pkg/daemon/daemon.go:372]: Linter complained.

File: pkg/daemon/daemon.go (hunk 7 above)
Comment (DmytroLinkin) [reply pkg/daemon/daemon.go:389]: same as previous

File: pkg/daemon/daemon.go (hunk 8 above)
Comment (DmytroLinkin) [reply pkg/daemon/daemon.go:554]: Same reason as for subnet manager. Probably we can let ib-k8s gona fail here without backoff as on manager validation to notify user about configuration issues, but not sure how this will impact on UX.

File: pkg/daemon/daemon.go (hunk 9 above)
Comment (DmytroLinkin) [reply pkg/daemon/daemon.go:412]: Ok, If U asking

File: pkg/daemon/daemon.go (hunk 10 above)
Comment (DmytroLinkin) [reply pkg/daemon/daemon.go:362]: > in this case we will retry this network in the next periodic update call right ?

Right

> why would we want to do that ? in case of corrupt net-attach-def object ?

I assume for cases, when error wasn't persistent.

File: pkg/daemon/daemon.go (hunk 9 above)
Comment (adrianchiris) [reply pkg/daemon/daemon.go:412]: I prefer we have it as i suggested

File: pkg/daemon/daemon.go (hunk 8 above)
Comment (adrianchiris) [reply pkg/daemon/daemon.go:554]: OK lets keep it as is

File: pkg/daemon/daemon.go (hunk 4 above)
Comment (adrianchiris) [reply pkg/daemon/daemon.go:185]: ok

File: pkg/daemon/daemon.go (hunk 3 above)
Comment (adrianchiris) [reply pkg/daemon/daemon.go:119]: OK lets add warning to the log in case of failure when retrying like other cases

File: pkg/daemon/daemon.go (hunk 2 above)
Comment (adrianchiris) [reply pkg/daemon/daemon.go:60]: i would suggest a reword:
 ```
In this case, there may be an increased execution time as <api call time> may take longer as retries are performed by k8s client.
k8s client backoff time is determined by KUBE_CLIENT_BACKOFF_BASE, KUBE_CLIENT_BACKOFF_DURATION environment variables [1]......
```

**Comment by adrianchiris** [comment]
One concern about the delete not rescheduling is that it may leave the SM with stale entries if/when it comes back online.
In any case this type of cleaning should be handled in a separate routine.

Im LGTM on these changes.
